File: src/main/scala/com/ajjpj/amapper/core/impl/AMapperWorkerImpl.scala
package com.ajjpj.amapper.core.impl

import com.ajjpj.amapper.core._
import scala.collection.mutable.ArrayBuffer

/**
 * @author arno
 */
private[impl] class AMapperWorkerImpl[H] (valueMappings: CanHandleSourceAndTargetCache[AValueMappingDef[_,_, _ >: H]],
                                     objectMappings: CanHandleSourceAndTargetCache[AObjectMappingDef[_,_, _ >: H]],
                                     val logger: AMapperLogger,
                                     val helpers: H, val identifierExtractor: IdentifierExtractor,
                                     contextExtractor: AContextExtractor, preProcessor: CanHandleSourceAndTargetCache[APreProcessor], postProcessor: CanHandleSourceAndTargetCache[APostProcessor],
                                     deferredWork: ArrayBuffer[()=>Unit]) extends AMapperWorker[H] {
  private val identityCache = new IdentityCache
  val diffBuilder = new ADiffBuilder

  override def map(path: PathBuilder, source: AnyRef, target: AnyRef, types: QualifiedSourceAndTargetType, context: Map[String, AnyRef]) = {
    valueMappings.entryFor(types) match {
      case Some(v) => Some(mapValue(path, source, types, context))
      case None => mapObject(path, source, target, types, context)
    }
  }

  override def mapValue(path: PathBuilder, source: AnyRef, types: QualifiedSourceAndTargetType, context: Map[String, AnyRef]) = {
    logger.debug ("map: " + source + " @ " + path.build)
    val m = valueMappingFor(types, path)
    m.asInstanceOf[AValueMappingDef[AnyRef, AnyRef, H]].map(source, types, this, context)
  }

  private def valueMappingFor(types: QualifiedSourceAndTargetType, path: PathBuilder) = valueMappings.entryFor(types) match {
    case Some(m) => m.asInstanceOf[AValueMappingDef[AnyRef, AnyRef, H]]
    case None => throw new AMapperException("no value mapping def found for " + types, path.build)
  }

  private def objectMappingFor(types: QualifiedSourceAndTargetType, path: PathBuilder) = objectMappings.entryFor(types) match {
    case Some(m) => m.asInstanceOf[AObjectMappingDef[AnyRef, AnyRef, H]]
    case None => throw new AMapperException("no object mapping def found for " + types, path.build)
  }


  override def mapObject(path: PathBuilder, sourceRaw: AnyRef, target: AnyRef, types: QualifiedSourceAndTargetType, context: Map[String, AnyRef]) = {
    logger.debug ("map: " + sourceRaw + " @ " + path.build)

    val m = objectMappingFor(types, path)
    val preProcessed = preProcessor
      .entryFor(types)
      .map(_.preProcess(sourceRaw, types)) // apply preprocessor (if any)
      .getOrElse(Some(sourceRaw)) // no preprocessor: leave 'as is'

    preProcessed.map(source => {
      val newContext = contextExtractor.withContext(context, sourceRaw, types.sourceType)
      val resultRaw = m.asInstanceOf[AObjectMappingDef[AnyRef, AnyRef, H]].map(source, target, types, this, newContext, path)

      val result = postProcessor
        .entryFor(types)
        .map(_.postProcess(resultRaw, types))
        .getOrElse (resultRaw)

      if(m.isCacheable)
        identityCache.register(source, result, path)

      result
    })
  }

  /**
   * Deferred mapping causes this mapping to be deferred until no non-deferred mapping work is left. The underlying
   *  abstraction is that there is a primary, hierarchical structure through which all objects can be
   *  reached, and a secondary structure of non-containment references. <p />
   * The former is mapped first so as to provide well-defined, 'normalized' paths to every element. Therefore, every
   *  object must be reachable by <em>exactly one</em> non-deferred path; ambiguities are not permitted and must be
   *  resolved by making one of the references 'deferred'.<p />
   * All objects must be reachable through the primary hierarchy. Put differently, it is invalid for
   *  an unmapped object to be reached through a deferred 'map' call. <p />
   * NB: deferred mapping automatically means mutable state in the target object structure because the result of the
   *  mapping is created after the initial object structure is complete. This also prevents streaming processing.
   */
  override def mapDeferred(path: PathBuilder, sourceRaw: AnyRef, target: => AnyRef, types: QualifiedSourceAndTargetType, callback: (AnyRef) => Unit) {
    logger.debug ("map deferred: " + types + " @ " + path.build)
    deferredWork += (() => {
      logger.debug("processing deferred: " + types + " @ " + path.build)
      val source = sourceRaw //TODO deProxyStrategy(sourceRaw)
      identityCache.get(source) match {
        case Some(prevTarget) =>
          callback(prevTarget)
        case None =>
          logger.deferredWithoutInitial(path.build) //TODO special treatment for collections etc. --> flag in the mapping def?
        // create a new, empty context: context is accumulated only from parents to children
        mapObject(path, source, target, types, Map[String, AnyRef]()) match {
          case Some(mapped) => callback(mapped)
          case _ =>
        }
      }
    })
  }

  override def diff(path: PathBuilder, sourceOld: AnyRef, sourceNew: AnyRef, types: QualifiedSourceAndTargetType, oldContext: Map[String, AnyRef], newContext: Map[String, AnyRef], isDerived: Boolean) {
    valueMappings.entryFor(types) match {
      case Some(v) => diffValue (path, sourceOld, sourceNew, types, oldContext, newContext, isDerived)
      case None    => diffObject(path, sourceOld, sourceNew, types, oldContext, newContext, isDerived)
    }
  }

  override def diffValue(path: PathBuilder, sourceOld: AnyRef, sourceNew: AnyRef, types: QualifiedSourceAndTargetType, oldContext: Map[String, AnyRef], newContext: Map[String, AnyRef], isDerived: Boolean) {
    logger.debug ("diff: " + sourceOld + " <-> " + sourceNew + " @ " + path.build)
    valueMappingFor(types, path).diff(diffBuilder, sourceOld, sourceNew, types, this, oldContext, newContext, path, isDerived)
  }

  override def diffObject(path: PathBuilder, sourceOldRaw: AnyRef, sourceNewRaw: AnyRef, types: QualifiedSourceAndTargetType, contextOld: Map[String, AnyRef], contextNew: Map[String, AnyRef], isDerived: Boolean) {
    logger.debug ("diff: " + sourceOldRaw + " <-> " + sourceNewRaw + " @ " + path.build)
    val pre = preProcessor.entryFor(types)
    val preProcessedOld: Option[AnyRef] = pre
      .map(_.preProcess(sourceOldRaw, types)) // apply preprocessor (if any)
      .getOrElse(Some(sourceOldRaw)) // no preprocessor: leave 'as is'
    val preProcessedNew: Option[AnyRef] = pre
      .map(_.preProcess(sourceNewRaw, types)) // apply preprocessor (if any)
      .getOrElse(Some(sourceNewRaw)) // no preprocessor: leave 'as is'

    (preProcessedOld.isDefined, preProcessedNew.isDefined) match {
      case (false, false) =>
      case (true, true) =>
        val sourceOld = preProcessedOld.get
        val sourceNew = preProcessedNew.get

        doDiffObject(path, sourceOld, sourceNew, types, contextOld, contextNew, isDerived)
      case _ =>
        logger.diffPreProcessMismatch(path.build)
    }
  }

  private def doDiffObject(path: PathBuilder, sourceOld: AnyRef, sourceNew: AnyRef, types: QualifiedSourceAndTargetType, contextOldOrig: Map[String, AnyRef], contextNewOrig: Map[String, AnyRef], isDerived: Boolean) {
    val oldContext = contextExtractor.withContext (contextOldOrig, sourceOld, types.sourceType)
    val newContext = contextExtractor.withContext (contextNewOrig, sourceNew, types.sourceType)

    var causesDerived = false
    if(sourceOld == null && sourceNew != null) {
      diffBuilder.add (AddDiffElement (path.build, identifierExtractor.uniqueIdentifier(sourceNew, types.sourceType), isDerived))
      causesDerived = true
    }
    else if(sourceOld != null && sourceNew == null) {
      diffBuilder.add (RemoveDiffElement (path.build, identifierExtractor.uniqueIdentifier(sourceOld, types.sourceType), isDerived))
      causesDerived = true
    }
    else {
      val oldIdent = identifierExtractor.uniqueIdentifier (sourceOld, types.sourceType)
      val newIdent = identifierExtractor.uniqueIdentifier (sourceNew, types.sourceType)
      if (oldIdent != newIdent) {
        diffBuilder.add (ChangeRefDiffElement (path.build, oldIdent, newIdent, isDerived))
        causesDerived = true
      }
    }

    objectMappingFor(types, path).diff (diffBuilder, sourceOld, sourceNew, types, this, oldContext, newContext, path, isDerived || causesDerived)
    identityCache.register((sourceOld, sourceNew), this, path)
  }

  override def diffDeferred(path: PathBuilder, sourceOldRaw: AnyRef, sourceNewRaw: AnyRef, types: QualifiedSourceAndTargetType, contextOld: Map[String, AnyRef], contextNew: Map[String, AnyRef], isDerived: Boolean) {
    logger.debug ("diff deferred: " + sourceOldRaw + " <-> " + sourceNewRaw + " @ " + path.build)
    deferredWork += (() => {
      logger.debug("processing deferred diff: " + types + "@" + path.build)

      val pre = preProcessor.entryFor(types)
      val preProcessedOld: Option[AnyRef] = pre
        .map(_.preProcess(sourceOldRaw, types)) // apply preprocessor (if any)
        .getOrElse(Some(sourceOldRaw)) // no preprocessor: leave 'as is'
      val preProcessedNew: Option[AnyRef] = pre
          .map(_.preProcess(sourceNewRaw, types)) // apply preprocessor (if any)
          .getOrElse(Some(sourceNewRaw)) // no preprocessor: leave 'as is'

      (preProcessedOld.isDefined, preProcessedNew.isDefined) match {
        case (false, false) =>
        case (true, true) if identityCache.get((preProcessedOld, preProcessedNew)).isDefined =>
        case (true, true) =>
          logger.deferredWithoutInitial(path.build)

          doDiffObject(path, preProcessedOld.get, preProcessedNew.get, types, contextOld, contextNew, isDerived)
        case _ =>
          logger.diffPreProcessMismatch(path.build)
      }
    })
  }
}

File: src/main/java/com/ajjpj/amapper/core/impl/AMapperWorkerImpl.java
package com.ajjpj.amapper.core.impl;

import com.ajjpj.abase.collection.immutable.AHashMap;
import com.ajjpj.abase.collection.immutable.AMap;
import com.ajjpj.abase.collection.immutable.AOption;
import com.ajjpj.abase.function.AFunction0;
import com.ajjpj.abase.function.AFunction0NoThrow;
import com.ajjpj.abase.function.AStatement0;
import com.ajjpj.abase.function.AStatement1;
import com.ajjpj.amapper.core.*;
import com.ajjpj.amapper.core.exclog.AMapperExceptionHandler;
import com.ajjpj.amapper.core.exclog.AMapperLogger;
import com.ajjpj.amapper.core.path.APath;
import com.ajjpj.amapper.core.tpe.AQualifiedSourceAndTargetType;
import com.ajjpj.amapper.core.tpe.CanHandleSourceAndTargetCache;

import java.util.Queue;

/**
 * @author arno
 */
public class AMapperWorkerImpl<H> implements AMapperWorker<H> {
    private final CanHandleSourceAndTargetCache<AValueMappingDef<?,?,? super H>, AValueMappingDef<Object, Object, H>> valueMappings;
    private final CanHandleSourceAndTargetCache<AObjectMappingDef<?,?,? super H>, AObjectMappingDef<Object, Object, H>> objectMappings;
    private final AMapperLogger logger;
    private final H helpers;
    private final AIdentifierExtractor identifierExtractor;
    private final AContextExtractor contextExtractor;
    private final CanHandleSourceAndTargetCache<APreProcessor, APreProcessor> preProcessors;
    private final CanHandleSourceAndTargetCache<APostProcessor, APostProcessor> postProcessors;
    private final Queue<AStatement0<RuntimeException>> deferredWork;

    private final IdentityCache identityCache = new IdentityCache();

    public AMapperWorkerImpl(CanHandleSourceAndTargetCache<AValueMappingDef<?, ?, ? super H>, AValueMappingDef<Object, Object, H>> valueMappings,
                             CanHandleSourceAndTargetCache<AObjectMappingDef<?, ?, ? super H>, AObjectMappingDef<Object, Object, H>> objectMappings,
                             AMapperLogger logger, H helpers,
                             AIdentifierExtractor identifierExtractor, AContextExtractor contextExtractor,
                             CanHandleSourceAndTargetCache<APreProcessor, APreProcessor> preProcessor,
                             CanHandleSourceAndTargetCache<APostProcessor, APostProcessor> postProcessor,
                             Queue<AStatement0<RuntimeException>> deferredWork) {
        this.valueMappings = valueMappings;
        this.objectMappings = objectMappings;
        this.logger = logger;
        this.helpers = helpers;
        this.identifierExtractor = identifierExtractor;
        this.contextExtractor = contextExtractor;
        this.preProcessors = preProcessor;
        this.postProcessors = postProcessor;
        this.deferredWork = deferredWork;
    }

    @Override public AMapperLogger getLogger() {
        return logger;
    }

    @Override public H getHelpers() {
        return helpers;
    }

    @Override public AIdentifierExtractor getIdentifierExtractor() {
        return identifierExtractor;
    }

    @Override public AOption<Object> map(APath path, Object source, Object target, AQualifiedSourceAndTargetType types, AMap<String, Object> context) {
        try {
            final AOption<AValueMappingDef<Object, Object, H>> vm = valueMappings.tryEntryFor(types);
            if(vm.isDefined()) {
                return AOption.some(vm.get().map(source, types, this, context));
            }
            else {
                return mapObject(path, source, target, types, context);
            }
        } catch (Exception exc) {
            return AMapperExceptionHandler.onError(exc, path);
        }
    }

    @Override public AOption<Object> mapObject(final APath path, final Object sourceRaw, Object target, AQualifiedSourceAndTargetType types, AMap<String, Object> context) {
        logger.debug (new AFunction0NoThrow<String>() {
            @Override
            public String apply() {
                return "map: " + sourceRaw + " @ " + path;
            }
        });

        try {
            final AObjectMappingDef<Object, Object, H> m = objectMappings.expectedEntryFor(types, path);
            final AOption<Object> preProcessed = preProcess(preProcessors.allEntriesFor(types), sourceRaw, types);

            if(preProcessed.isEmpty()) {
                // preprocessor vetoed this mapping operation
                return AOption.none();
            }

            final Object source = preProcessed.get();

            final AOption<Object> cached = identityCache.get (source, types.target ());
            if(cached.isDefined()) {
                return cached;
            }

            final AMap<String, Object> newContext = contextExtractor.withContext(context, sourceRaw, types.sourceType);
            final Object resultRaw = m.map(source, target, types, this, newContext, path);

            final Object result = postProcess(postProcessors.allEntriesFor(types), resultRaw, types);

            if(m.isCacheable()) {
                identityCache.register(source, result, types.target (), path);
            }

            return AOption.some(result);
        } catch (Exception exc) {
            return AMapperExceptionHandler.onError(exc, path);
        }
    }

    private AOption<Object> preProcess(Iterable<APreProcessor> allPre, Object sourceRaw, AQualifiedSourceAndTargetType types) {
        AOption<Object> result = AOption.some(sourceRaw);
        for(APreProcessor pre: allPre) {
            result = pre.preProcess(result.get(), types);
            if(result.isEmpty()) {
                return result;
            }
        }
        return result;
    }

    private Object postProcess(Iterable<APostProcessor> allPost, Object sourceRaw, AQualifiedSourceAndTargetType types) {
        Object curResult = sourceRaw;
        for(APostProcessor post: allPost) {
            curResult = post.postProcess(curResult, types);
        }
        return curResult;
    }

    @Override public Object mapValue(final APath path, final Object source, AQualifiedSourceAndTargetType types, AMap<String, Object> context) {
        logger.debug (new AFunction0NoThrow<String>() {
            @Override
            public String apply() {
                return "map: " + source + " @ " + path;
            }
        });

        try {
            final AValueMappingDef<Object, Object, H> vm = valueMappings.expectedEntryFor(types, path);
            return vm.map(source, types, this, context);
        }
        catch(Exception exc) {
            return AMapperExceptionHandler.onError(exc, path);
        }
    }

    @Override public void mapDeferred(final APath path, final Object sourceRaw, final AFunction0<Object, Exception> target, final AQualifiedSourceAndTargetType types, final AStatement1<Object, Exception> callback) {
        logger.debug (new AFunction0NoThrow<String>() {
            @Override public String apply() {
                return "map deferred: " + types + " @ " + path;
            }
        });

        deferredWork.add(new AStatement0<RuntimeException>() {
            @Override public void apply() {
                logger.debug(new AFunction0NoThrow<String>() {
                    @Override
                    public String apply() {
                        return "processing deferred: " + types + " @ " + path;
                    }
                });

                try {
                    final AOption<Object> preProcessed = preProcess(preProcessors.allEntriesFor(types), sourceRaw, types);

                    if(preProcessed.isEmpty()) {
                        return;
                    }
                    final Object source = preProcessed.get();

                    final AOption<Object> prevTarget = identityCache.get(source, types.target ());
                    if(prevTarget.isDefined()) {
                        callback.apply(prevTarget.get());
                    }
                    else {
                        logger.deferredWithoutInitial(path); //TODO special treatment for collections etc. --> flag in the mapping def?
                        // create a new, empty context: context is accumulated only from parents to children
                        final AOption<Object> mapped = mapObject(path, source, target.apply (), types, AHashMap.<String, Object>empty());
                        if(mapped.isDefined()) {
                            callback.apply(mapped.get());
                        }
                        //TODO documentation: explain why there is no post processing done here
                    }
                }
                catch (Exception exc) {
                    AMapperExceptionHandler.onError(exc, path);
                }
            }
        });
    }
}
